File: leetcode/0020-valid-parentheses/solution.cpp
#include<iostream>
#include<stack>
#include<unordered_map>
#include<string>

using namespace std;

class Solution {
public:
    bool isValid(const string& s) {
        unordered_map < char , char > map = {{'(',')'},{'{','}'},{'[',']'}};
        stack<char>stack;
        for(char c : s)
        {
            if(map.find(c) != map.end())
            {
                // if the character is an opening bracket,push it onto the stack
                stack.push(c);
            }
            else
            {
                // IF the stack is empty or the top of the stack doesn't match the closing bracket
                if(stack.empty() || map[stack.top()] !=c)
                {
                    return false;
                }
                // Pop the matching opening bracket
                stack.pop();
            }
        }

        // If the stack is empty, all the brackets are properly closed
        return stack.empty();
    }

};

File: leetcode/0020-valid-parentheses/solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public boolean isValid(String s) {
        
        if((s.length()&1)==1) return false;
        
        Deque<Character> st = new ArrayDeque<>();
        
        for(char c : s.toCharArray()){
            
            if(c=='['){
                
                st.push(']');
            }
            else if(c=='{'){
                
                st.push('}');
            }
            else if(c=='('){
                
                st.push(')');
            }
            else{
                
                if(st.isEmpty() || st.pop() != c) return false;
            }
        }
        return st.isEmpty();
    }
}
